### PR TITLE
Comparison export labels

### PR DESCRIPTION
**Pull request by egor-ryashin**
## Summary by CodeRabbit

New Features:
- Enhanced the `MetricsViewComparison` functionality with an `export` parameter, enabling users to generate and export label columns for data.
- Updated the `generalExport` function to handle the newly introduced label columns, improving data export capabilities.

Refactor:
- Modified the `buildMetricsTopListSQL` and `buildMetricsComparisonTopListSQL` functions in the `MetricsViewComparison` struct to support the new features, ensuring seamless integration and functionality.

### CODE REVIEW

File: runtime/queries/metricsview_comparison_toplist.go [hunk 1]
```diff
@@ -237,8 +245,10 @@ func (q *MetricsViewComparison) buildMetricsTopListSQL(mv *runtimev1.MetricsView
 				return "", nil, err
 			}
 			selectCols = append(selectCols, fmt.Sprintf("%s as %s", expr, safeName(m.Name)))
+			labelCols = append(labelCols, fmt.Sprintf("%s as %s", safeName(m.Name), labelMap[m.Name]))
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:248]: Doesn't it need `safeName` for the label? (The label can have arbitrary characters, including spaces and quotes)

File: runtime/queries/metricsview_comparison_toplist.go [hunk 2]
```diff
@@ -237,8 +245,10 @@ func (q *MetricsViewComparison) buildMetricsTopListSQL(mv *runtimev1.MetricsView
...
 		case runtimev1.BuiltinMeasure_BUILTIN_MEASURE_COUNT:
 			selectCols = append(selectCols, fmt.Sprintf("COUNT(*) as %s", safeName(m.Name)))
+			labelCols = append(labelCols, fmt.Sprintf("%s as %s", safeName(m.Name), labelMap[m.Name]))
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:251]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 3]
```diff
@@ -248,6 +258,7 @@ func (q *MetricsViewComparison) buildMetricsTopListSQL(mv *runtimev1.MetricsView
 				return "", nil, fmt.Errorf("builtin measure '%s' expects non-empty string argument, got '%v'", m.BuiltinMeasure.String(), m.BuiltinMeasureArgs[0])
 			}
 			selectCols = append(selectCols, fmt.Sprintf("COUNT(DISTINCT %s) as %s", safeName(arg), safeName(m.Name)))
+			labelCols = append(labelCols, fmt.Sprintf("%s as %s", safeName(m.Name), labelMap[m.Name]))
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:261]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 4]
```diff
@@ -361,20 +397,41 @@ func (q *MetricsViewComparison) buildMetricsComparisonTopListSQL(mv *runtimev1.M
 				safeName(m.Name+"__delta_abs"),
 				safeName(m.Name+"__delta_rel"),
 			)
+			labelTuple = fmt.Sprintf(
+				"base.%[1]s AS %[5]s, comparison.%[1]s AS %[2]s, base.%[1]s - comparison.%[1]s AS %[3]s, (base.%[1]s - comparison.%[1]s)/comparison.%[1]s::DOUBLE AS %[4]s",
+				safeName(m.Name),
+				labelMap[m.Name]+" (prev)",
+				labelMap[m.Name]+" (Δ)",
+				labelMap[m.Name]+" (Δ%)",
+				labelMap[m.Name],
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:406]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 5]
```diff
@@ -361,20 +397,41 @@ func (q *MetricsViewComparison) buildMetricsComparisonTopListSQL(mv *runtimev1.M
...
 				"ANY_VALUE(base.%[1]s), ANY_VALUE(comparison.%[1]s), ANY_VALUE(base.%[1]s - comparison.%[1]s), ANY_VALUE(SAFE_DIVIDE(base.%[1]s - comparison.%[1]s, CAST(comparison.%[1]s AS DOUBLE)))",
 				safeName(m.Name),
 			)
+			labelTuple = fmt.Sprintf(
+				"ANY_VALUE(base.%[1]s) AS %[2]s, ANY_VALUE(comparison.%[1]s) AS %[3]s, ANY_VALUE(base.%[1]s - comparison.%[1]s) AS %[4]s, ANY_VALUE(SAFE_DIVIDE(base.%[1]s - comparison.%[1]s, CAST(comparison.%[1]s AS DOUBLE))) AS %[5]s",
+				safeName(m.Name),
+				labelMap[m.Name],
+				labelMap[m.Name]+" (prev)",
+				labelMap[m.Name]+" (Δ)",
+				labelMap[m.Name]+" (Δ%)",
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:419]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 6]
```diff
@@ -708,25 +773,25 @@ func (q *MetricsViewComparison) generalExport(ctx context.Context, rt *runtime.R
 	meta[0] = &runtimev1.MetricsViewColumn{
 		Name: q.DimensionName,
 	}
-	if !isTimeRangeNil(q.ComparisonTimeRange) {
+	if comparison {
 		for i, m := range q.Result.Rows[0].MeasureValues {
 			meta[1+i*4] = &runtimev1.MetricsViewColumn{
-				Name: m.MeasureName,
+				Name: measureMap[m.MeasureName],
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:779]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 7]
```diff
@@ -698,8 +755,16 @@ func (q *MetricsViewComparison) generalExport(ctx context.Context, rt *runtime.R
 		}
 	}
 
+	measureMap := make(map[string]string, len(mv.Measures))
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:758]: Call it `labelMap` like elsewhere?

File: runtime/queries/metricsview_comparison_toplist.go [hunk 8]
```diff
@@ -708,25 +773,25 @@ func (q *MetricsViewComparison) generalExport(ctx context.Context, rt *runtime.R
...
-				Name: fmt.Sprintf("%s__previous", m.MeasureName),
+				Name: fmt.Sprintf("%s (prev)", measureMap[m.MeasureName]),
 			}
 			meta[3+i*4] = &runtimev1.MetricsViewColumn{
-				Name: fmt.Sprintf("%s__delta_abs", m.MeasureName),
+				Name: fmt.Sprintf("%s (Δ)", measureMap[m.MeasureName]),
 			}
 			meta[4+i*4] = &runtimev1.MetricsViewColumn{
-				Name: fmt.Sprintf("%s__delta_rel", m.MeasureName),
+				Name: fmt.Sprintf("%s (Δ%%)", measureMap[m.MeasureName]),
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:788]: Same

File: runtime/queries/metricsview_comparison_toplist.go [hunk 9]
```diff
@@ -708,25 +773,25 @@ func (q *MetricsViewComparison) generalExport(ctx context.Context, rt *runtime.R
...
 			}
 		}
 	} else {
 		for i, m := range q.Result.Rows[0].MeasureValues {
 			meta[1+i] = &runtimev1.MetricsViewColumn{
-				Name: m.MeasureName,
+				Name: measureMap[m.MeasureName],
```
Comment (begelundmuller) [inline runtime/queries/metricsview_comparison_toplist.go:794]: Same

**Review by begelundmuller — CHANGES_REQUESTED** [review]

**Review by begelundmuller — CHANGES_REQUESTED** [review]
Did an experience pass – it works well for measure labels, but seems like handling for the dimension label is still missing:

Screenshot 2023-11-07 at 13 56 50

**Comment by egor-ryashin** [comment]
image
Smoke test

**Review by begelundmuller — APPROVED** [review]

**Pull request merged**